File: Java-Core/src/main/java/presto/expression/FetchExpression.java
package presto.expression;

import java.util.ArrayList;
import java.util.List;

import presto.error.NullReferenceError;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.grammar.Identifier;
import presto.parser.Section;
import presto.runtime.Context;
import presto.runtime.Variable;
import presto.type.BooleanType;
import presto.type.CollectionType;
import presto.type.IType;
import presto.type.ListType;
import presto.utils.CodeWriter;
import presto.value.Boolean;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.ListValue;

public class FetchExpression extends Section implements IExpression {

	Identifier itemName;
	IExpression source;
	IExpression filter;
	
	public FetchExpression(Identifier itemName, IExpression source, IExpression filter) {
		this.itemName = itemName;
		this.source = source;
		this.filter = filter;
	}
	

	@Override
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
			writer.append("fetch any ");
			writer.append(itemName);
			break;
		case O:
			writer.append("fetch (");
			writer.append(itemName);
			writer.append(")");
			break;
		case S:
			writer.append("fetch ");
			writer.append(itemName);
			break;
		}
		writer.append(" from ");
		source.toDialect(writer);
		writer.append(" where ");
		filter.toDialect(writer);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType sourceType = source.check(context);
		if(!(sourceType instanceof CollectionType))
			throw new SyntaxError("Expecting a list, set or tuple as data source !");
		Context local = context.newLocalContext();
		IType itemType = ((CollectionType)sourceType).getItemType();
		local.registerValue(new Variable(itemName, itemType));
		IType filterType = filter.check(local);
		if(filterType!=BooleanType.instance())
			throw new SyntaxError("Filtering expresion must return a boolean !");
		return new ListType(itemType);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		Object src = source.interpret(context);
		if(src==null)
			throw new NullReferenceError();
		if(!(src instanceof ICollection<?>))
			throw new InternalError("Illegal fetch source: " + source);
		IType sourceType = source.check(context);
		if(!(sourceType instanceof CollectionType))
			throw new InternalError("Illegal source type: " + sourceType.getName());
		IType itemType = ((CollectionType)sourceType).getItemType();
		List<IValue> result = new ArrayList<IValue>();
		Context local = context.newLocalContext();
		Variable item = new Variable(itemName, itemType);
		local.registerValue(item);
		for(IValue o : ((ICollection<?>)src).getItems(context)) {
			local.setValue(itemName, o);
			Object test = filter.interpret(local);
			if(!(test instanceof Boolean))
				throw new InternalError("Illegal test result: " + test);
			if(((Boolean)test).getValue())
				result.add(o);
		}
		return new ListValue(itemType, result);
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/expression/FetchExpression.js
var Section = require("../parser/Section").Section;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var BooleanType = require("../type/BooleanType").BooleanType;
var ListType = require("../type/ListType").ListType;
var TupleType = require("../type/TupleType").TupleType;
var SetType = require("../type/SetType").SetType;
var TransientVariable = require("../runtime/TransientVariable").TransientVariable;
var NullReferenceError = require("../error/NullReferenceError").NullReferenceError;
var ListValue = require("../value/ListValue").ListValue;
var TupleValue = require("../value/TupleValue").TupleValue;
var SetValue = require("../value/SetValue").SetValue;
var Bool = require("../value/Bool").Bool;

function FetchExpression(itemName, source, filter) {
	Section.call(this);
	this.itemName = itemName;
	this.source = source;
	this.filter = filter;
	return this;
}

FetchExpression.prototype  = Object.create(Section.prototype);
FetchExpression.prototype.constructor = FetchExpression;


FetchExpression.prototype.toDialect = function(dialect) {
	return "fetch any " + this.itemName + " from " + this.source.toString() + " where " + this.filter.toString();
};

FetchExpression.prototype.check = function(context) {
	var listType = this.source.check(context);
	if(!(listType instanceof ListType || listType instanceof TupleType || listType instanceof SetType)) {
		throw new SyntaxError("Expecting a list type as data source !");
	}
	var local = context.newLocalContext();
	local.registerValue(new TransientVariable(this.itemName,listType.itemType));
	var filterType = this.filter.check(local);
	if(filterType!=BooleanType.instance) {
		throw new SyntaxError("Filtering expresion must return a boolean !");
	}
	return listType;
};

FetchExpression.prototype.interpret = function(context) {
	var listType = this.source.check(context);
	if(!(listType instanceof ListType || listType instanceof TupleType || listType instanceof SetType)) {
		throw new InternalError("Illegal source type: " + listType.getName());
	}
	var list = this.source.interpret(context);
	if(list==null) {
		throw new NullReferenceError();
	}
	if(!(list instanceof ListValue || list instanceof TupleValue || list instanceof SetValue)) {
		throw new InternalError("Illegal fetch source: " + this.source);
	}
	var itemType = listType.itemType;
	var result = new ListValue(itemType);
	var local = context.newLocalContext();
	var item = new TransientVariable(this.itemName, itemType);
	local.registerValue(item);
	var iter = list.getIterator(context);
	while(iter.hasNext()) {
		var o = iter.next();
		local.setValue(this.itemName, o);
		var test = this.filter.interpret(local);
		if(!(test instanceof Bool)) {
			throw new InternalError("Illegal test result: " + test);
		}
		if(test.value) {
			result.add(o);
		}
	}
	return result;
};

FetchExpression.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

FetchExpression.prototype.toEDialect = function(writer) {
    writer.append("fetch any ");
    writer.append(this.itemName);
    writer.append(" from ");
    this.source.toDialect(writer);
    writer.append(" where ");
    this.filter.toDialect(writer);
};


FetchExpression.prototype.toODialect = function(writer) {
    writer.append("fetch (");
    writer.append(this.itemName);
    writer.append(")");
    writer.append(" from ");
    this.source.toDialect(writer);
    writer.append(" where ");
    this.filter.toDialect(writer);
};

FetchExpression.prototype.toSDialect = function(writer) {
    writer.append("fetch ");
    writer.append(this.itemName);
    writer.append(" from ");
    this.source.toDialect(writer);
    writer.append(" where ");
    this.filter.toDialect(writer);
};

exports.FetchExpression = FetchExpression;
